src/services/factories/getPriceEstimation.ts

```diff
@@ -0,0 +1,59 @@
+import BigNumber from 'bignumber.js'
+import { assert } from '@gnosis.pm/dex-js'
+
+import { ZERO_BIG_NUMBER } from 'const'
+
+import { TheGraphApi } from 'api/thegraph/TheGraphApi'
+import { TokenList } from 'api/tokenList/TokenListApi'
+
+export interface GetPriceParams {
+  networkId: number
+  numeratorTokenId: number
+  denominatorTokenId: number
+}
+
+export function getPriceEstimationFactory(factoryParams: {
+  theGraphApi: TheGraphApi
+  tokenListApi: TokenList
+}): (params: GetPriceParams) => Promise<BigNumber> {
+  const { theGraphApi, tokenListApi } = factoryParams
+
+  return async (params: GetPriceParams): Promise<BigNumber> => {
+    const { networkId, numeratorTokenId, denominatorTokenId } = params
+
+    assert(numeratorTokenId !== denominatorTokenId, 'Token ids cannot be the equal')
+
+    // ---
+    // For an 'optimization' of sorts, we always query all known tokens plus the requested token ids.
+    // Assuming the caching is on for the TheGraphApi layer (it was last time I checked) and we don't
+    // expect to have tokens ids not often outside of our list.
+    // Thus, we can actually make a request once, and reuse the response for all token pairs falling into the same list
+
+    // all known token ids
+    const tokenIdsSet = new Set(tokenListApi.getTokens(networkId).map(token => token.id))
+    // just in case the tokens are not in our initial list:
+    tokenIdsSet.add(numeratorTokenId)
+    tokenIdsSet.add(denominatorTokenId)
+    // make it nice and sorted to help with caching
+    const tokenIds = [...tokenIdsSet].sort()
```

Comment:
_For an 'optimization' of sorts :wink:_ can presort in parent scope  if numeratorTokenId and denominatorTokenId are in the set, like they will usually be, and will certainly be included on repeated calls
Like:
```ts
// L20:
const tokenIdsSet = new Set(tokenListApi.getTokens(networkId).map(token => token.id))
const tokenIds = Array.from(tokenIdsSet).sort() // just my preference over ...spread for single iterables, fell free to ignore
...
// L33:
let needToResort = false
if (needToResort = needToResort || !tokenIdsSet.has(numeratorTokenId)) {
  tokenIdsSet.add(numeratorTokenId)
  
  tokenIds.push(numeratorTokenId)
}
if (needToResort = needToResort || !tokenIdsSet.has(denominatorTokenId)) {
  tokenIdsSet.add(denominatorTokenId)
  
  tokenIds.push(denominatorTokenId)
}
if (needToResort) {
  tokenIds.sort() // it sorts in place
}
```

_optimization intensifies_ :feelsgood:

Reply:
1 little problem with your logic, but it's a problem that I'm willing to be authoritarian about anyway.

It requires the `networkId` on factory creation time. 

But if we really think about it, doesn't make sense to have prices for any network other than Mainnet.

1. That's where we have trades and actual prices can be derived
2. For now only the mainnet endpoint (staging) can be queried for prices :D

So, I'll enforce the networkId and we are good :)

Reply:
https://github.com/gnosis/dex-react/pull/733/commits/eb3c84047e656bfa13ad4d1c9ee332b82a55c4f1

Reply:
Props to you for switching statement order in `needToResort || !tokenIdsSet.has(denominatorTokenId) -> addTokenId(denominatorTokenId) || needsToSort`